aws-lc-rs/src/rsa.rs

```diff
@@ -232,54 +235,41 @@ impl RsaKeyPair {
         signature: &mut [u8],
     ) -> Result<(), Unspecified> {
         let encoding = padding_alg.encoding();
-        let mut output_len = self.public_modulus_len();
-        if signature.len() != output_len {
+
+        let mut md_ctx = digest::digest_ctx::DigestContext::new_uninit();
+        let mut pctx = MaybeUninit::<*mut EVP_PKEY_CTX>::uninit();
+        let digest = digest::match_digest_type(&encoding.0.id);
+
+        if 1 != unsafe {
+            EVP_DigestSignInit(
+                md_ctx.as_mut_ptr(),
+                pctx.as_mut_ptr(),
+                *digest,
+                null_mut(),
+                *self.rsa_key,
+            )
+        } {
             return Err(Unspecified);
         }
-        unsafe {
-            let digest_alg = encoding.0;
-            let digest = digest::digest(digest_alg, msg);
-            let digest = digest.as_ref();
-
-            let padding = encoding.1;
-            // These functions are non-mutating of RSA:
-            // https://github.com/awslabs/aws-lc/blob/main/include/openssl/rsa.h#L286
-            let result = match padding {
-                RsaPadding::RSA_PKCS1_PADDING => {
-                    let mut output_len = c_uint::try_from(output_len)?;
-                    let digest_len = digest.len();
-                    let result = RSA_sign(
-                        digest_alg.hash_nid,
-                        digest.as_ptr(),
-                        digest_len,
-                        signature.as_mut_ptr(),
-                        &mut output_len,
-                        *self.rsa_key,
-                    );
-                    debug_assert_eq!(output_len as usize, signature.len());
-                    result
-                }
-                RsaPadding::RSA_PKCS1_PSS_PADDING => {
-                    let result = RSA_sign_pss_mgf1(
-                        *self.rsa_key,
-                        &mut output_len,
-                        signature.as_mut_ptr(),
-                        output_len,
-                        digest.as_ptr(),
-                        digest.len(),
-                        *match_digest_type(&digest_alg.id),
-                        null(),
-                        -1,
-                    );
-                    debug_assert_eq!(output_len, signature.len());
-                    result
-                }
-            };
 
-            if result != 1 {
+        if let RsaPadding::RSA_PKCS1_PSS_PADDING = encoding.1 {
+            let pctx = unsafe { pctx.assume_init() };
+            if 1 != unsafe { EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_PSS_PADDING) } {
                 return Err(Unspecified);
-            }
+            };
+            if 1 != unsafe { EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, -1) } {
```

Comment:
`RSA_PSS_SALTLEN_DIGEST` might help clear things up with the confusing API design :(
```suggestion
            if 1 != unsafe { EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, RSA_PSS_SALTLEN_DIGEST) } {
```